Support scanning by hostname

dav3r opened the pull request:
<!-- GitHub renders PRs such that soft line breaks are treated as hard
line breaks.  In order to make this template render as expected we
therefore have to avoid soft breaks and therefore will offend
markdownlint with long lines.  This is the reason for the markdownline
disable directive just below.

For more details see:
https://github.com/github/markup/issues/1050#issuecomment-294654762 -->
<!-- markdownlint-disable MD013 -->
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR adds support in the CyHy database for defining hostnames to be scanned, not just IP addresses.

It also adds two new scripts:
* `cyhy-domain`: Similar to the existing `cyhy-ip`, this script allows an operator to add/remove/move hostnames from the inventory of a CyHy entity.
* `cyhy-domainsync`: This is a new script that is meant to be run on a regular schedule.  It is responsible for "housekeeping" tasks related to hostnames in CyHy, namely:
  * Creating HostDocs for hostnames that resolve to IP addresses that weren't already defined in CyHy inventory
  * Updating HostDocs to track hostnames that resolve to the same IP, but have different CyHy owners
  * Removing outdated hostname info from HostDocs (i.e. if a hostname no longer resolves to an IP address that it used to)
  * Removing HostDocs that are no longer needed because they no longer have any valid hostnames

Additionally, the `pytest` tests were updated to make them all pass again (see https://github.com/cisagov/cyhy-core/pull/101/commits/8965376f8dcf06a786b6fe073853111775804200 and https://github.com/cisagov/cyhy-core/pull/101/commits/843497e8436c37f89705c168686bedf7f87bf050).

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

This is a feature that has long been requested and now we are finally able to implement it.

This PR will resolve (partially) https://github.com/cisagov/cyhy-system/issues/70, in conjunction with:
* https://github.com/cisagov/cyhy-commander/pull/14
* https://github.com/cisagov/ansible-role-cyhy-commander/pull/88
* https://github.com/cisagov/cyhy_amis/pull/859
* https://github.com/cisagov/cyhy-reports/pull/128
* https://github.com/cisagov/ncats-webd/pull/40

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

* [x] Built a CyHy database AMI with the changes in this PR
* [x] Deployed a fresh database instance based on that AMI in a test environment
* [x] Confirmed that the instance and cyhy-commander started up correctly
* [x] Imported request doc JSON for a test entity
* [x] Verified that `cyhy-domain` functionality worked as expected
* [x] Verified that `cyhy-domainsync` functionality worked as expected

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.
- [x] Bump major, minor, patch, pre-release, and/or build versions [as appropriate](https://semver.org/#semantic-versioning-specification-semver) via the `bump_version` script *if* this repository is versioned *and* the changes in this PR [warrant a version bump](https://semver.org/#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api).

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Create a release (necessary if and only if the version was bumped).
- [x] Build fresh production AMI
- [x] Deploy updated database instance to production
- [x] Ensure [updated TicketDoc indices](https://github.com/cisagov/cyhy-core/pull/101/commits/f1653c27a11bc9b878188c979e4b52255ba395e7) are applied in Production
- [x] Work with CyHy team to ensure they can create an updated Docker image and use it

dav3r committed:
Sort RequestDoc keys alphabetically

dav3r committed:
Sort HostDoc keys alphabetically

dav3r committed:
Sort VulnScanDoc keys alphabetically

dav3r committed:
Sort TicketDoc keys alphabetically

dav3r committed:
Add hostnames/hostname to structure of some collections

dav3r committed:
Add hostname field to relevant TicketDoc indices

dav3r committed:
Make fetch_ready_hosts return HostDocs instead of IPs

This allows the commander to create vulnscan jobs that include the hostname(s) in the HostDoc, rather than just the IP.

dav3r committed:
Add a missing comma

dav3r committed:
Include hostnames dict (if present) when grabbing hosts in get_some_for_stage()

dav3r committed:
Use optional hostname when determining IP/host owner

dav3r committed:
Consider hostname when opening/reopening/updating tickets

dav3r committed:
Add comment explaining why hostname is not needed in ticket-closing pipeline query

When we scan an IP address, that scan includes all hostnames associated with that IP.  This means we can close all matching tickets (IP/port/source_id/source) that weren't seen during the scan (i.e. not in self.__seen_ticket_ids).

dav3r committed:
Add null hostname to legacy testing vuln docs

This enables the existing tests to work with the updated (hostname-aware) code.

dav3r committed:
Remove an old, commented-out, unused line

dav3r committed:
Add tests for hostname-aware ticket manager

dav3r committed:
Add validators package to requirements

It is used by the new cyhy-domain script.

dav3r committed:
Add new cyhy-domain script

This script is used to manage domain (hostname) ownership similar to how cyhy-ip is used to manage IP address ownership in Cyber Hygiene.

felddy committed:
Merge changes to setup.py

felddy committed:
Add first version of cyhy-domainsync

dav3r committed:
Change HostDoc hostnames field from dict to list of dicts

dav3r committed:
Add get_by_hostname function to HostDoc

It returns all HostDocs with the given hostname in their list of hostnames.

felddy committed:
Add support for config file argument

felddy committed:
DRY out code by calling HostDoc functions

felddy committed:
Remove unused imports

felddy committed:
Improve log messages when deleting HostDocs

dav3r committed:
Remove hostnames from HostDocs as part of domain removal process

dav3r committed:
Pin validators package to 0.14.2

That is the last version that correctly installs for Python 2.7.x.

felddy committed:
Add argument for default owner

felddy committed:
Close tickets when hostname does not resolve

felddy committed:
Improve logging message format and clarity

felddy committed:
Get collection reference from object

dav3r committed:
Make error output consistent

dav3r committed:
Clean up and simplify command usage section

dav3r committed:
Look up starting stage for default owner

New hosts created for the default owner will be set to the correct starting stage.

dav3r committed:
Include IP geolocation when creating HostDocs for default owner

dav3r committed:
Update domainsync logic to handle case where hostname owner has changed

jsf9k commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+def do_list_all(db):
+    requests = db.RequestDoc.find({"hostnames": {"$exists": True}}).sort("_id")
+    for r in requests:
+        print "# %s (%s): %d" % (
+            r["agency"]["name"],
+            r["_id"],
+            len(r["hostnames"])
+        )
```
I'm confused as to why this script uses Python2-style `print` (`print` is not a function) whereas `cyhy-domainsync` uses Python3-style `print` (`print` is a function).  Both files have the same shebang so they should be running the same version of Python.

dav3r replied:
This is because `cyhy-domainsync` was written from scratch, whereas `cyhy-domain` is based on the much older `cyhy-ip` script, which still has not yet been converted to Python 3 syntax.

jsf9k replied:
OK, but the non-function syntax doesn't work in Python 3 and the function syntax doesn't work in Python 2.  Both scripts have the same shebang, so they are presumably running in the same Python interpreter.  It seems that one of them must fail just due to the `print` syntax.

dav3r replied:
The function syntax does work in Python 2, which is what we have in the environment where these scripts are currently being used.

jsf9k reviewed (commented):
I have one question.

dav3r committed:
Bump version from 0.0.2 to 1.0.0

dav3r committed:
Define a default owner

This owner's request document will be created when the system initializes.  Initial use is to own all "ownerless" HostDocs, which are created when a CyHy entity has a hostname that resolves to an IP address not already owned by a CyHy entity.  This is how we account for cases where an entity owns a hostname, but not necessarily the IP addresses that it resolves to.

dav3r committed:
Add new HostDoc utility function to get the owner IDs of all hostnames associated with an IP address

dav3r committed:
Consider hostname when opening/reopening/updating tickets based on portscans

dav3r committed:
Add HostDoc.get_hosts_with_hostnames

It returns HostDocs whose list of hostnames contains any of the given list of hostnames.

dav3r committed:
Include hostnames in SnapshotDoc

dav3r committed:
Make hosts collection queries include hostname owners

dav3r committed:
Include hostname hosts in list of networks included in snapshots

These changes ensure that the hosts will be included in the snapshot, even if they are technically owned by the default CYHY owner.

jsf9k commented on cyhy/db/chdatabase.py:
```diff
@@ -467,13 +466,29 @@ def create_snapshot(
+                    for host in self.__db.HostDoc.get_hosts_with_hostnames(
+                        descendant_doc["hostnames"]
+                    ):
+                        snapshot_networks.add(host.ip)
```
This is duplicated code, so it would be nice to create a function for this and use it here and just above.

dav3r replied:
Excellent suggestion, please see https://github.com/cisagov/cyhy-core/pull/101/commits/c20eeb147d5d8952b2856308ab3b41e583050ca3.

jsf9k reviewed (approved):
I have one minor suggestion, which you can take or leave.  Otherwise, this all looks good to me.

dav3r committed:
Remove unused imports from test_host_state_manager.py

dav3r committed:
Fix a typo and update an import path in test_nmap_handler.py

dav3r committed:
Update host state manager tests to account for an additional output value

It's been many years since the finished_stage output value was added to the transition function in host_state_manager, but we never updated the tests to account for it.

dav3r commented on cyhy/db/chdatabase.py:
```diff
@@ -467,13 +466,29 @@ def create_snapshot(
+        snapshot_networks = netaddr.IPSet()
+        if owner_doc:
+            snapshot_networks = owner_doc.networks
+            snapshot_doc["hostnames"] = owner_doc.get("hostnames", [])
```
This is not an issue because when a fresh SnapshotDoc is created, the `hostnames` field is automatically set to an empty list:

```python
> db.SnapshotDoc()["hostnames"]
> []
```

jsf9k replied:
I suggest adding a comment here with the explanation you gave as to why this is not an issue.

dav3r replied:
Good idea, stay tuned.  You'll see a code comment included once I'm done testing the code changes for https://github.com/cisagov/cyhy-core/pull/101#discussion_r2175247373.

dav3r replied:
You can see the code comment in https://github.com/cisagov/cyhy-core/pull/101/commits/c20eeb147d5d8952b2856308ab3b41e583050ca3.

dav3r commented on cyhy/db/chdatabase.py:
```diff
@@ -273,16 +273,15 @@ def check_host_next_scans(self):
-            int_ip = host["_id"]
-            self.transition_host(int_ip)
-            ips.append(host["ip"])
-        return ips
+            self.transition_host(host["_id"])
+        return hosts
```
This is correct- I chose to update all callers to handle full host documents because this function is only called in one place.  See https://github.com/cisagov/cyhy-commander/pull/14/commits/1f120c1cd258334096bd8e7e2a74e8fe8a16fb32 for where this was done.

jsf9k reviewed (approved):
_Approval intensifies!!!_

dav3r committed:
DRY out some repeated code in create_snapshot()

Co-authored-by: Shane Frasier <jeremy.frasier@gwe.cisa.dhs.gov>

dav3r committed:
Bump version of cyhy-domainsync from 0.0.1 to 1.0.0

felddy reviewed (approved):
`v1.0.0` 🎉 
I'm thankful for your generous comments.  They'll pay dividends in the future... if it ever comes.

mcdonnnj commented on cyhy/db/chdatabase.py:
```diff
@@ -467,13 +466,31 @@ def create_snapshot(
+        def add_networks_and_hostnames_to_snapshot(request_doc, snapshot_doc):
+            """Adds networks and hostnames from a request document to a snapshot
+               document. If the request document has hostnames, it also adds the
+               IP addresses of those hostnames to the snapshot."""
```
Nitpick but would you reformat this to follow standard Python function comment style? That is a single line title, empty line, and then additional explanation.

dav3r replied:
Please see https://github.com/cisagov/cyhy-core/pull/101/commits/6b78888c7820254a20b75d1111a90372af7b621e.

mcdonnnj commented on cyhy/db/chdatabase.py:
```diff
@@ -467,13 +466,31 @@ def create_snapshot(
+                        snapshot_doc["networks"].add(host.ip)
+
+        owner_doc = self.__db.RequestDoc.get_by_owner(owner)
+        snapshot_doc["networks"] = netaddr.IPSet()
```
If we cannot use the [`SnapshotDoc.networks` property](https://github.com/cisagov/cyhy-core/blob/5661d725e8ca192948234fe7dfc4e3ad02519ac7/cyhy/db/database.py#L1581-L1583) and we must change the field type before converting it back then I would like some documentation around that process.

dav3r replied:
I'm not quite following you here.  The [`SnapshotDoc.networks` property](https://github.com/cisagov/cyhy-core/blob/5661d725e8ca192948234fe7dfc4e3ad02519ac7/cyhy/db/database.py#L1581-L1583) is used to get the networks from an existing snapshot document.  In this particular section of code, we are creating a new snapshot and populating it, so I'm not clear on how you are wanting me to use `SnapshotDoc.networks` since `snapshot_doc["networks"]` has not been assigned a value until the line you mentioned above: https://github.com/cisagov/cyhy-core/blob/f932d087351070dc7ea3196f6af30616e23278e7/cyhy/db/chdatabase.py#L487

dav3r replied:
@mcdonnnj Can you please circle back to this conversation?

mcdonnnj replied:
So if you generate a new snapshot with `SnapshotDoc()` the value of `SnapshotDoc["networks"]` is an empty list and accessing the property with `SnapshotDoc.networks` returns `IPSet([])`. It doesn't help in this case anyway so let me reframe the primary concern in what I was saying :

The assignment `snapshot_doc["networks"] = netaddr.IPSet()` does not conform to the ODM https://github.com/cisagov/cyhy-core/blob/8af7e153cb9162283a7368204bd02d578109dd7c/cyhy/db/database.py#L1509 and the code at https://github.com/cisagov/cyhy-core/blob/8af7e153cb9162283a7368204bd02d578109dd7c/cyhy/db/chdatabase.py#L494 is necessary to re-align the document with the ODM. I would just like documentation around the type switching. This would help it makes more clear why, if you're referencing the ODM, the inner function `add_networks_and_hostnames_to_snapshot()` is performing a set operation on what you would expect to be a list.

dav3r replied:
I think I understand your concern now.  So would a comment like this suffice?
```python
        # Convert networks from an IPSet to a list of CIDRs to conform to the
        # SnapshotDoc schema.
        snapshot_doc["networks"] = snapshot_doc["networks"].iter_cidrs()
```

mcdonnnj replied:
I think I would also like a brief sibling comment at the assignment just mentioning that this it is temporary/for processing or similar. Just to cover both sides of the manipulation.

dav3r replied:
That's fair; please have a look at https://github.com/cisagov/cyhy-core/pull/101/commits/0ae82f55bbd813854914d1589209a6d11e084adc.

mcdonnnj reviewed (commented):
Just dropping the feedback I already had so you can address it while I get back on top of reviewing the PRs around this effort.

dav3r committed:
Use standard Python function comment style

Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+        sys.exit(0)
+
+    if args.get("OWNER"):
+        if not db.RequestDoc.find_one({"_id": args["OWNER"]}):
```
Is there a reason we can't do this?

```suggestion
        if not db.RequestDoc.get_by_owner(args["OWNER"]):
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+            sys.exit(-1)
+    
+    if args.get("NEW_OWNER"):
+        if not db.RequestDoc.find_one({"_id": args["NEW_OWNER"]}):
```
Same question.

```suggestion
        if not db.RequestDoc.get_by_owner(args["NEW_OWNER"]):
```

mcdonnnj commented on bin/cyhy-domainsync:
```diff
@@ -0,0 +1,195 @@
+        geo_loc_db = GeoLocDB()
+
+        # Look up starting stage for default owner
+        default_owner_req = db.RequestDoc.find_one({"_id": default_owner})
```
Another.

```suggestion
        default_owner_req = db.RequestDoc.get_by_owner(default_owner)
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+
+
+def do_list(db, owner):
+    request = db.RequestDoc.find_one({"_id": owner})
```
Another.

```suggestion
    request = db.RequestDoc.get_by_owner(owner)
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+    if domain_conflict:
+        sys.exit(-1)
+
+    request = db.RequestDoc.find_one({"_id": owner})
```
Another.

```suggestion
    request = db.RequestDoc.get_by_owner(owner)
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+
+
+def remove(db, owner, domains):
+    request = db.RequestDoc.find_one({"_id": owner})
```
Another.

```suggestion
    request = db.RequestDoc.get_by_owner(owner)
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+def do_list(db, owner):
+    request = db.RequestDoc.find_one({"_id": owner})
+    for h in request.get("hostnames", []):
+        print(h)
```
Consistently use Py2 style `print` statements.

```suggestion
        print h
```

dav3r replied:
Since we want to move this code to Py3, I don't see any reason to not use syntax that is acceptable in both Py2 and Py3.  It's just less stuff for us to change later.

mcdonnnj replied:
It's mainly about consistency in this file. We have both Py2 and Py3 statements scattered throughout the file.  Ideally we would just slap in `from __future__ import print_function` and enforce the Py3-style, but without that I'd at least prefer that in a given file it's consistent which is used.

dav3r replied:
This is not on my list of things worth quibbling about right now, so I'm going to say that this is fine as is and can be cleaned up later when we move to Py3.

jsf9k replied:
I'm OK with that.

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+            len(r["hostnames"])
+        )
+        for h in sorted(r["hostnames"]):
+            print("\t%s" % h)
```
Consistency again.

```suggestion
            print "\t%s" % h
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+    if invalid_domains:
+        print "ERROR - invalid domain(s):"
+        for d in invalid_domains:
+            print(d)
```
Consistency.

```suggestion
            print d
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+
+    print "Domains added to %s (%s):" % (request["agency"]["name"], owner)
+    for d in domains:
+        print("\t%s" % d)
```
Consistency.

```suggestion
        print "\t%s" % d
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+            request["agency"]["name"], owner
+        )
+        for d in sorted(unowned_domains):
+            print("\t%s" % d)
```
Consistency.

```suggestion
            print "\t%s" % d
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+
+    print "Domains removed from %s (%s):" % (request["agency"]["name"], owner)
+    for d in domains:
+        print("\t%s" % d)
```
Consistency.

```suggestion
        print "\t%s" % d
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+            domains = matched_domains[owner]
+            print "# %s: %s" % (owner, len(domains))
+            for d in sorted(domains):
+                print("\t%s" % d)
```
Consistency.

```suggestion
                print "\t%s" % d
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+    if unmatched_domains:
+        print "# Unmatched: %s" % len(unmatched_domains)
+        for d in sorted(unmatched_domains):
+            print("\t%s" % d)
```
Consistency.

```suggestion
            print "\t%s" % d
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+            orig_owner_request["agency"]["name"]
+        )
+        for d in domains_not_in_orig_owner:
+            print("\t%s" % d)
```
Consistency.

```suggestion
            print "\t%s" % d
```

mcdonnnj commented on bin/cyhy-domain:
```diff
@@ -0,0 +1,294 @@
+        new_owner_request["agency"]["name"], new_owner
+        )
+    for d in sorted(domains_to_move):
+        print("\t%s" % d)
```
Consistency.

```suggestion
        print "\t%s" % d
```

mcdonnnj commented on bin/cyhy-domainsync:
```diff
@@ -0,0 +1,195 @@
+  --version                       Show version.
+"""
+
+# standard python libraries
```
```suggestion
# Standard Python Libraries
```

mcdonnnj commented on bin/cyhy-domainsync:
```diff
@@ -0,0 +1,195 @@
+# standard python libraries
+import logging
+
+# third-party libraries (install with pip)
```
```suggestion
# Third-Party Libraries
```

mcdonnnj commented on bin/cyhy-domainsync:
```diff
@@ -0,0 +1,195 @@
+from docopt import docopt
+from netaddr import IPAddress, AddrFormatError
+
+# intra-project modules
```
```suggestion
# cisagov Libraries
```

mcdonnnj commented on bin/cyhy-domainsync:
```diff
@@ -0,0 +1,195 @@
+            resolver.default_resolver.nameservers = [args["--resolver"]]
+
+        default_owner = args["--default-owner"]
+        print("Using default owner: " + default_owner)
```
I'm not sure if you want to use Py2 `print` formatting here. Everything is consistently using Py3 within the script so you could go either way.

dav3r replied:
It doesn't make a difference to me either way, so I'm fine as is.

mcdonnnj commented on bin/cyhy-domainsync:
```diff
@@ -0,0 +1,195 @@
+                    # ]
+
+                    # If the hostnames list doesn't exist, create it
+                    if not host_doc.get("hostnames"):
```
I think we're specifically looking for a missing hostnames list but this would match empty ones as well. Not a big deal but not really inline with the documentation.
```suggestion
                    if host_doc.get("hostnames") is None:
```

mcdonnnj commented on bin/cyhy-domainsync:
```diff
@@ -0,0 +1,195 @@
+
+                    # If the hostname isn't already in the list, add it
+                    if not found_hostname:
+                        host_doc["hostnames"].append({"hostname": hostname, "owner": r["_id"]})
```
Would this be better served eliminating the `found_hostname` variable and simply doing a `for`/`else`?

dav3r replied:
Good thinking - please see https://github.com/cisagov/cyhy-core/pull/101/commits/08c5b1c988cc5d83baea1e4e5f45d9d715a0c96b.

dav3r committed:
Consistently use RequestDoc.get_by_owner() instead of find_one()

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Use our standard import section header comments

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Create hostnames field in HostDoc if it doesn't exist

There's no need to do anything if it already exists and it is an empty list.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Get rid a temp variable by using a for/else loop

Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add explanatory comments about the networks field during snapshot creation

Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>
